### DISCUSSION TITLE
remote.meta.sync dir glob-patterns

### DISCUSSION THREAD

**Discussion by Material-Scientist** [seaweedfs/seaweedfs#5898] [Ideas]
Here's an example of a very large S3 bucket (~5TB with ~20M files):
![image](https://github.com/user-attachments/assets/e75ac547-16cb-43f9-9c10-589a26e39867)

It takes ~2-3 days just to mount it, and it's similarly slow to synchronize it all with `remote.meta.sync -dir=/buckets/<bucket>`

Sure, I can set up a cron job to sync only the folders of the current date, but as you can see those are very nested.

Would it instead be possible to allow defining glob-patterns, such as:
```
remote.meta.sync -dir=/buckets/binance-vision/data/spot/daily/klines/*/*/*-2020-06-03.zip*
```

That way, it hopefully only takes minutes to sync, instead of hours or days.

**Comment by chrislusf** [comment]
The bottleneck is on S3 side. According to https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html 
* each request can only get up-to 1000 entries
* the only way to narrow down the listing scope is by `prefix` and `start-after`.
So the initial mount would have to take that long, if there are many files.

The incremental scanning may have some opportunities to optimize.
My current idea is to find the last entry in a folder, and set it to `start-after`.

But not sure how to handle deletions. It would only work for finding new entries. 
We can make a `remote.meta.sync -dir=/buckets/xxx -additionOnlyMode`

**Reply by chrislusf** [reply]
The right way to solve this would be to setup AWS S3 file change events https://aws.amazon.com/blogs/aws/s3-event-notification/ and publish it to SQS, and then write some code to subscribe to SQS, and process the the event in SeaweedFS.

**Reply by Material-Scientist** [reply]
`-additionOnlyMode` would already be a great start!
In the buckets I'm mounting, all old files are immutable.
